bots/tests-scan

```diff
@@ -366,21 +351,16 @@ def update_status(api, revision, context, last, changes):
         return False
     return True
 
-def cockpit_tasks(api, update, contexts, repo, pull_data, pull_number, sha, amqp):
+def cockpit_tasks(api, update, branch_contexts, repo, pull_data, pull_number, sha, amqp, is_external):
     results = []
-    branch_contexts = { }
-    for (context, branches) in contexts.items():
-        for branch in branches:
-            if branch not in branch_contexts:
-                branch_contexts[branch] = [ ]
-            branch_contexts[branch].append(context)
-
     pulls = []
+    contexts = set(itertools.chain(*branch_contexts.values()))
+
     if pull_number or pull_data:
         # Avoid processing pull requests with the same number on external
         # repositories: if repo is None, the api is querying the repo the pull
         # request belongs to
-        if repo is None:
+        if not is_external:
```

Comment:
It's rather hard to see why this weird special case needs to happen here. It's the sole reason this new `is_external` parameter has to be passed through, and it's not obvious why scanning for a particular open PR (and *only* that!) would need to know about this fact.

So let's say we use this to scan cockpit-podman PR #123, with `tests-scan --repo cockpit-project/cockpit-podman -p 123`. Why would we not want this code to run here? How would it collide with cockpit PR #123? `api` should be initialized for podman here.

It's likely that I just forgot the details of this. Stuff like this is what makes tests-scan so insanely hard to maintain.. @Gundersanne , do you happen to remember?

Reply:
I think the use case should be somehow different. This behavior is (current upstream, ignoring this PR for now)  reached only when none repo is passed (assuming we are at c-p/c) see [here](https://github.com/cockpit-project/cockpit/blob/master/bots/tests-scan#L158) so therefore two options come to mind (both from c-p/c checkout)

- `tests-scan`
- `tests-scan -p 123`

For both of these options `scan_external_projects` would be executed and `scan_for_pull_tasks` is executed for all external projects which is pretty much calling `cockpit_tasks`. For the first case it picks up all open PR and does its magic, for the second case it does nothing (because pull_number is defined but repo is not None).

> So let's say we use this to scan cockpit-podman PR #123, with tests-scan --repo cockpit-project/cockpit-podman -p 123

It would not go past [this](https://github.com/cockpit-project/cockpit/blob/master/bots/tests-scan#L158), so this is not the usecase

Maybe I misunderstood but it seems to me that we want to `scan_external_projects` always when we scan all PRs in cockpit-project/cockpit. Is there use-case to run `tasks-scan` without specifying repo or PR and expecting it to scan all PRs is all projects? If so, then we can drop that condition in `cockpit_task` and go with `if not opts.repo and not (opts.pull_number or opts.pull_data): scan_external_projects()` in `main`. Or somehow accordingly.

btw. it was introduced in `333d7d0a53` by @Gundersanne

Reply:
So originally this line was there because `scan_external_projects` would get called when a cockpit PR was being scanned (so `bots/tests-scan -p 5000` for instance). And it would look for a PR with the same number on external projects. Now because cockpit PR numbers are so much higher this might not ever have become a problem practically.

In order to not break behaviour with the way `bots/tests-scan` worked previously, this was added. 

Because external projects each have their own webhook anway, I think `scan_external_projects` might only be needed when calling `bots/tests-scan` without `-p`. Rather than trying to ignore the `-p` for external projects in cockpit_tasks.

Reply:
If `-p` is specified (or `--pull-data`, which ought to be pretty much equivalent) should have no influence at all on whether a project/repo/context is considered "external". However, it *should*  disable `scan_external_projects`. This is a nasty hack right now, originating from the fact that we don't let projects declare their test-scan invocations themselves. This is related to:

> it seems to me that we want to scan_external_projects always when we scan all PRs in cockpit-project/cockpit

This applies to this piece of code:
```py
        # for the default (Cockpit) repo, also check external project PRs
        if not opts.repo:
            results += scan_external_projects(opts)
```

This is the bolted-on hack, and it's also wrong -- if `opts.repo == 'cockpit-project/cockpit`, we also want this to happen. However, I think it would be cleaner and easier to understand to completely drop this snippet, replace it with an `--all-external` option, and add `tests-scan --all-external` to cockpit's `.tasks`,  with a comment that in the future this ought to get split out into the external projects themselves. Even cleaner would be to literally do this right now -- cockpituous now cycles through all project's .tasks, so adding a `tests-scan` invocation to these should be (almost?) possible now. @mvollmer looked into that a few months ago, but it got stuck on something that I forgot.

> Is there use-case to run tasks-scan without specifying repo or PR and expecting it to scan all PRs is all projects

See above. We currently expect this use  case, but IMHO it would be better to make this explicit - either with a separate call and a new option, or calling it from different project's .tasks.

But calling tests-scan without `--repo` should continue to work -- it should then scan for the git checkout origin's repo, as it does now.